client/mm/mm_arb_market_maker.go

```diff
@@ -0,0 +1,651 @@
+	var matchID order.MatchID
+	copy(matchID[:], match.MatchID)
+
+	a.matchesMtx.Lock()
+	if _, seen := a.matchesSeen[matchID]; seen {
+		a.matchesMtx.Unlock()
+		return
+	}
+	a.matchesSeen[matchID] = true
+	a.matchesMtx.Unlock()
+
+	var cexRate uint64
+	if o.Sell {
+		cexRate = uint64(float64(match.Rate) / (1 + a.cfg.Profit))
+	} else {
+		cexRate = uint64(float64(match.Rate) * (1 + a.cfg.Profit))
+	}
+	cexRate = steppedRate(cexRate, a.mkt.RateStep)
+
+	a.cexTradesMtx.Lock()
+	defer a.cexTradesMtx.Unlock()
+
+	tradeID, err := a.cex.Trade(a.ctx, dex.BipIDSymbol(a.base), dex.BipIDSymbol(a.quote), !o.Sell, cexRate, match.Qty, a.cexTradeUpdatesID)
+	if err != nil {
+		a.log.Errorf("Error sending trade to CEX: %v", err)
+		return
+	}
+
+	// Keep track of the epoch in which the trade was sent to the CEX. This way
+	// the bot can cancel the trade if it is not filled after a certain number
+	// of epochs.
+	a.cexTrades[tradeID] = a.currEpoch.Load()
+}
+
+func (a *arbMarketMaker) processDEXMatchNote(note *core.MatchNote) {
+	var oid order.OrderID
+	copy(oid[:], note.OrderID)
+
+	a.ordMtx.RLock()
+	o, found := a.ords[oid]
+	if !found {
+		a.ordMtx.RUnlock()
+		return
+	}
+	a.ordMtx.RUnlock()
+
+	a.processDEXMatch(o, note.Match)
+}
+
+func (a *arbMarketMaker) processDEXOrderNote(note *core.OrderNote) {
+	var oid order.OrderID
+	copy(oid[:], note.Order.ID)
+
+	a.ordMtx.Lock()
+	o, found := a.ords[oid]
+	if !found {
+		a.ordMtx.Unlock()
+		return
+	}
+	a.ords[oid] = note.Order
+	a.ordMtx.Unlock()
+
+	for _, match := range note.Order.Matches {
+		a.processDEXMatch(o, match)
+	}
+
+	if !note.Order.Status.IsActive() {
+		a.ordMtx.Lock()
+		delete(a.ords, oid)
+		delete(a.oidToPlacement, oid)
+		a.ordMtx.Unlock()
+	}
+}
+
+func (a *arbMarketMaker) vwap(sell bool, qty uint64) (vwap, extrema uint64, filled bool, err error) {
+	return a.cex.VWAP(dex.BipIDSymbol(a.base), dex.BipIDSymbol(a.quote), sell, qty)
+}
+
+type arbMMRebalancer interface {
+	vwap(sell bool, qty uint64) (vwap, extrema uint64, filled bool, err error)
+	groupedOrders() (buys, sells map[int][]*groupedOrder)
+}
+
+func (a *arbMarketMaker) placeMultiTrade(placements []*rateLots, sell bool) {
+	qtyRates := make([]*core.QtyRate, 0, len(placements))
+	for _, p := range placements {
+		qtyRates = append(qtyRates, &core.QtyRate{
+			Qty:  p.lots * a.mkt.LotSize,
+			Rate: p.rate,
+		})
+	}
+
+	var options map[string]string
+	if sell {
+		options = a.cfg.BaseOptions
+	} else {
+		options = a.cfg.QuoteOptions
+	}
+
+	orders, err := a.core.MultiTrade(nil, &core.MultiTradeForm{
+		Host:       a.host,
+		Sell:       sell,
+		Base:       a.base,
+		Quote:      a.quote,
+		Placements: qtyRates,
+		Options:    options,
+	})
+	if err != nil {
+		a.log.Errorf("Error placing rebalancing order: %v", err)
+		return
+	}
+
+	a.ordMtx.Lock()
+	for i, ord := range orders {
+		var oid order.OrderID
+		copy(oid[:], ord.ID)
+		a.ords[oid] = ord
+		a.oidToPlacement[oid] = placements[i].placementIndex
+	}
```

Comment:
MultiTrade in core can fail one trade and keep going, possibly causing this index to be off if an order is not appended.

Reply:
`MultiTrade` always returns the trades in the same order as they were passed in. If second one fails, then it won't try the third.